### PR TITLE
New Pull request for grep cheat sheet

### PR DESCRIPTION
**Pull request by adityabhushan**
- [ ] New Instant Answer
    - [x] Cheat Sheets: **`New Grep Cheat Sheet`**

Instant Answer Page: https://duck.co/ia/view/grep_cheat_sheet

[Maintainer](http://docs.duckduckhack.com/maintaining/guidelines.html): @adityabhushan

### CODE REVIEW

**Comment by daxtheduck** [comment]
@adityabhushan Hey!
        
Thanks for taking the time to contribute! We really appreciate it.

We work closely with every contributor to make Instant Answers the best they can be, so we appreciate your patience as we look over your code. From here, the process usually goes like this:

1. Pull Request is reviewed by the DuckDuckGo staff and community
2. Staff and community will leave feedback with any necessary updates to the function or design.
3. Once you've made any necessary corrections, then your Instant Answer will be merged and deployed live on DuckDuckGo!

If you have any questions along the way, feel free to ask them here. Our staff and community are also available on Slack to answer any questions you may have. If you'd like to join us there please head to https://quackslack.herokuapp.com/ to get an invite.

>More Info: http://docs.duckduckhack.com/submitting/submitting-overview.html

Thanks!

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

File: share/goodie/cheat_sheets/json/grep.json [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+    "name": "Grep",
+    "description": "A quick overview of using the 'grep' command in linux.",
+
+    "metadata": {
+        "sourceName": "",
+        "sourceUrl" : ""
+    },
+
+    "aliases": [
+        "grep", "grep command", "grep overview"
```
Comment (GuiltyDolphin) [inline share/goodie/cheat_sheets/json/grep.json]: Remove the first alias (it is covered by the cheat sheet's name).

**Comment by KAUTH** [comment]
I saw that the checks failed, from a quick look that I gave to the checks' details, you should not have in the "aliases" the same name as the cheat sheet, so no need to put "grep cheat sheet," also check some white spaces and a comma inside of the quotation marks.

File: share/goodie/cheat_sheets/json/grep.json [hunk 2]
```diff
@@ -0,0 +1,174 @@
+{
+    "id": "grep_cheat_sheet",
+    "name": "Grep",
+    "description": "A quick overview of using the 'grep' command in linux.",
+
+    "metadata": {
+        "sourceName": "",
+        "sourceUrl" : ""
```
Comment (tagawa) [inline share/goodie/cheat_sheets/json/grep.json]: Thanks for this cheat sheet @adityabhushan. We should ideally have a source, so please could you use the GNU site: https://www.gnu.org/software/grep/manual/grep.html
(or other reliable website if you prefer)

Comment (adityabhushan) [inline share/goodie/cheat_sheets/json/grep.json]: Thanks a lot, have modified the alias and  added the suggested source.

**Comment by daxtheduck** [comment]
Automated screenshot from [beta.duckduckgo.com](https://beta.duckduckgo.com/?q=grep cheat sheet)
                
                [Mobile screenshot](https://images.duckduckgo.com/iu/?u=https://ia-screenshots.s3.amazonaws.com/grep_cheat_sheet_mobile.png?DC0MahmaGS)

**Comment by adityabhushan** [comment]
Thanks a lot :-)

**Comment by tagawa** [comment]
@adityabhushan Cool, looks like it's working! Let's leave it up on the beta server for a few days to see if there's any feedback from the community: https://beta.duckduckgo.com/?q=grep+cheat+sheet&ia=cheatsheet&iax=1

**Comment by adityabhushan** [comment]
Sure Daniel. That seems like a great idea.

File: share/goodie/cheat_sheets/json/grep.json [hunk 3]
```diff
@@ -0,0 +1,174 @@
+{
+    "id": "grep_cheat_sheet",
+    "name": "Grep",
+    "description": "A quick overview of using the 'grep' command in linux.",
```
Comment (jophab) [inline share/goodie/cheat_sheets/json/grep.json]: Lets remove the period at end of description

File: share/goodie/cheat_sheets/json/grep.json [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+                "key": ".",
+                "val": "A single character"
+            },
+            {
+                "key": "*",
+                "val": "Zero or more occurances of previous character"
+            },
+            {
+                "key": "+",
+                "val": "One of more occurrence of the pattern. "
```
Comment (jophab) [inline share/goodie/cheat_sheets/json/grep.json]: Remove the period here

Comment (adityabhushan) [inline share/goodie/cheat_sheets/json/grep.json]: Hi, Removed periods from both places.

**Comment by daxtheduck** [comment]
Automated screenshot from [beta.duckduckgo.com](https://beta.duckduckgo.com/?q=grep cheat sheet)
                
                [Mobile screenshot](https://images.duckduckgo.com/iu/?u=https://ia-screenshots.s3.amazonaws.com/grep_cheat_sheet_mobile.png?2zPmMDngR8)

**Comment by daxtheduck** [comment]
Automated screenshot from [beta.duckduckgo.com](https://beta.duckduckgo.com/?q=grep cheat sheet)
                
                [Mobile screenshot](https://images.duckduckgo.com/iu/?u=https://ia-screenshots.s3.amazonaws.com/grep_cheat_sheet_mobile.png?11ntRXeXgD)

File: share/goodie/cheat_sheets/json/grep.json [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+            }
+        ],
+        "Options": [
+            {
+                "key": "-i",
+                "val": "'ignore' option - Ignores case for matching"
+            },
+            {
+                "key": "-v",
+                "val": "'inverse' option - Displays lines which dont match the expression"
```
Comment (tagawa) [inline share/goodie/cheat_sheets/json/grep.json]: Please change `dont` to `don't`

File: share/goodie/cheat_sheets/json/grep.json [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+            },
+            {
+                "key": "-n",
+                "val": "'number' option - Displays lines numbers along with lineswhich dont match the expression"
```
Comment (tagawa) [inline share/goodie/cheat_sheets/json/grep.json]: Please change:
`lines numbers` to `line numbers`
and
`lineswhich` to `lines which`
and
`dont` to `don't`

File: share/goodie/cheat_sheets/json/grep.json [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+                "key": "-e expression",
+                "val": "'expression' option - Specifies expression with this option. Can be used multiple times in a statement"
+            },
+            {
+                "key": "-f filename",
+                "val": "'file' option - Takes pattern from a file, one per line"
+            },
+            {
+                "key": "-E",
+                "val": "'Extended Regular Expression' option - Treates pattern as extended regular expression"
```
Comment (tagawa) [inline share/goodie/cheat_sheets/json/grep.json]: Please change `Treates` to `Treats`

File: share/goodie/cheat_sheets/json/grep.json [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+            }
+        ],
+        "RE Wildcards": [
+            {
+                "key": ".",
+                "val": "A single character"
+            },
+            {
+                "key": "*",
+                "val": "Zero or more occurances of previous character"
```
Comment (tagawa) [inline share/goodie/cheat_sheets/json/grep.json]: Please change `occurances` to `occurrences`

File: share/goodie/cheat_sheets/json/grep.json [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+                "key": "?",
+                "val": "Optional - Can occur zero or one times"
+            },
+            {
+                "key": ".",
+                "val": "A single character"
+            },
+            {
+                "key": "[xyz]",
+                "val": "Select any one of the character x, y or z"
```
Comment (tagawa) [inline share/goodie/cheat_sheets/json/grep.json]: Please change `character` to `characters`

File: share/goodie/cheat_sheets/json/grep.json [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+            },
+            {
+                "key": "[a-z]",
+                "val": "Select a single character within the ASCII range between 'a' and 'z'"
+            },
+            {
+                "key": "[^a-z]",
+                "val": "Select a single character with does not fall in the ASCII range between 'a' and 'z'"
```
Comment (tagawa) [inline share/goodie/cheat_sheets/json/grep.json]: Please change `with` to `which`

File: share/goodie/cheat_sheets/json/grep.json [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+            },
+            {
+                "key": "[:al­num:]",
+                "val": "All alphanumeric, equivalent to [a-zA-Z0-9]"
+            }
+        ],
+        "RE Anchors": [
+            {
+                "key": "^",
+                "val": "Begining of a string or line"
```
Comment (tagawa) [inline share/goodie/cheat_sheets/json/grep.json]: Please change `Begining` to `Beginning`

File: share/goodie/cheat_sheets/json/grep.json [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+                "key": "$",
+                "val": "End of a string or line"
+            },
+            {
+                "key": "^$",
+                "val": "Empty line"
+            },
+            {
+                "key": "\\<",
+                "val": "Begining of word"
```
Comment (tagawa) [inline share/goodie/cheat_sheets/json/grep.json]: Please change `Begining` to `Beginning`

File: share/goodie/cheat_sheets/json/grep.json [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+            }
+        ],
+        "Examples": [
+            {
+                "key": "grep 'first' example.txt",
+                "val": "Returns lines which have the word 'first' in the file example.txt"
+            },
+            {
+                "key": "grep -i 'first' example.txt",
+                "val": "Returns lines which have the word 'first' or 'First' or 'FIRST' in the file example.txt(returns resules which are case insensitive)"
```
Comment (tagawa) [inline share/goodie/cheat_sheets/json/grep.json]: Please change `resules` to `results`, and add a space before the first parenthesis (bracket).

File: share/goodie/cheat_sheets/json/grep.json [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+                "key": "grep -e 'first' -e 'second' example.txt",
+                "val": "Returns lines which have the word 'first' or 'second' in the file example.txt"
+            },
+            {
+                "key": "grep -l 'first' *.txt",
+                "val": "Returns list of all files with '.txt' as extension which contain the word 'first'"
+            },
+            {
+                "key": "grep 'fi*rst' example.txt",
+                "val": "Returns lines which have the word 'fist', 'first', 'firrst', 'firrrst' etc (with zero or more occunrrences of 'r') in the file example.txt"
```
Comment (tagawa) [inline share/goodie/cheat_sheets/json/grep.json]: Please change `occunrrences` to `occurrences`

**Comment by tagawa** [comment]
@adityabhushan Thank you for making the changes so quickly. I've just added a few more comments - sorry I didn't spot this before.

**Comment by adityabhushan** [comment]
Thanks :-)

Corrected all typos as suggested.

Thanks and Regards,
Aditya Bhushan Dwivedi
+91 72591 83906 | @adi_bhushan <https://twitter.com/adi_bhushan>

**Comment by daxtheduck** [comment]
Automated screenshot from [beta.duckduckgo.com](https://beta.duckduckgo.com/?q=grep cheat sheet)
                
                [Mobile screenshot](https://images.duckduckgo.com/iu/?u=https://ia-screenshots.s3.amazonaws.com/grep_cheat_sheet_mobile.png?h0EcIUywl3)

File: share/goodie/cheat_sheets/json/grep.json [hunk 15]
```diff
@@ -0,0 +1,174 @@
...
+            }
+        ],
+        "RE Wildcards": [
+            {
+                "key": ".",
+                "val": "A single character"
+            },
+            {
+                "key": "*",
+                "val": "Zero or more occurences of previous characters"
```
Comment (tagawa) [inline share/goodie/cheat_sheets/json/grep.json]: Please change `occurences` to `occurrences` (with 2 "r"s)

File: share/goodie/cheat_sheets/json/grep.json (hunk 9 above)
Comment (tagawa) [inline share/goodie/cheat_sheets/json/grep.json]: Please could you make this change too?

**Comment by daxtheduck** [comment]
Automated screenshot from [beta.duckduckgo.com](https://beta.duckduckgo.com/?q=grep cheat sheet)
                
                [Mobile screenshot](https://images.duckduckgo.com/iu/?u=https://ia-screenshots.s3.amazonaws.com/grep_cheat_sheet_mobile.png?w890LKPMcO)

**Comment by tagawa** [comment]
Thank you again @adityabhushan! Let's leave it open for a couple more days for any other review comments, then merge after that.

**Comment by adityabhushan** [comment]
Sure Daniel.

Thanks for being patient with typos.

Thanks and Regards,
Aditya Bhushan Dwivedi
+91 72591 83906 | @adi_bhushan <https://twitter.com/adi_bhushan>

**Comment by tagawa** [comment]
Heh, no problem! I do it all the time 😃

**Comment by tagawa** [comment]
Looks good to me and no further review comments, so let's merge 🎉 
Thank you @adityabhushan for creating this and to @jophab for the feedback (as always!).

And if you'd like to continue contributing, we're currently on a programming mission, aiming to improve Instant Answers for developers. More info on how to help out here: https://forum.duckduckhack.com/t/duckduckhack-programming-mission-overview/53

**Comment by daxtheduck** [comment]
@adityabhushan this is now merged! It should be live on DuckDuckGo within one week (or during the next release) and we'll be sure to notify you when it happens.

After your Instant Answer goes live for the World to use, you will be able to see [traffic information](http://docs.duckduckhack.com/maintaining/tools.html#monitor-your-instant-answer-page) about your IA as well as receive feedback directly from users. 
                
Please refer to the [maintenance guide and docs](http://docs.duckduckhack.com/maintaining/guidelines.html) to keep up with potential changes to your Instant Answer.
                
Thank you again for your help!

**Comment by adityabhushan** [comment]
Thanks a lot Davis, will keep contributing to the platform. :)

Thanks and Regards,
Aditya Bhushan Dwivedi
+91 72591 83906 | @adi_bhushan <https://twitter.com/adi_bhushan>

**Comment by tagawa** [comment]
@adityabhushan I'm pleased to say this is live now - thank you!
https://duckduckgo.com/?q=grep+cheat+sheet&ia=cheatsheet

**Pull request merged**